core/src/main/java/org/apache/iceberg/CachingCatalog.java

```diff
@@ -0,0 +1,100 @@
+import java.util.concurrent.atomic.AtomicBoolean;
+import org.apache.iceberg.catalog.Catalog;
+import org.apache.iceberg.catalog.TableIdentifier;
+import org.apache.iceberg.exceptions.AlreadyExistsException;
+
+public class CachingCatalog implements Catalog {
+
+  public static Catalog wrap(Catalog catalog) {
+    return new CachingCatalog(catalog);
+  }
+
+  private final Cache<TableIdentifier, Table> tableCache = Caffeine.newBuilder()
+      .softValues()
+      .expireAfterAccess(1, TimeUnit.MINUTES)
+      .build();
+  private final Catalog catalog;
+
+  private CachingCatalog(Catalog catalog) {
+    this.catalog = catalog;
+  }
+
+  @Override
+  public Table loadTable(TableIdentifier ident) {
+    return tableCache.get(ident, catalog::loadTable);
+  }
+
+  @Override
+  public Table createTable(TableIdentifier ident, Schema schema, PartitionSpec spec, String location,
+                           Map<String, String> properties) {
+    AtomicBoolean created = new AtomicBoolean(false);
+    Table table = tableCache.get(ident, identifier -> {
+      created.set(true);
+      return catalog.createTable(identifier, schema, spec, location, properties);
+    });
+
+    if (!created.get()) {
+      throw new AlreadyExistsException("Table already exists: %s", ident);
+    }
+
+    return table;
+  }
+
+  @Override
+  public Transaction newCreateTableTransaction(TableIdentifier ident, Schema schema, PartitionSpec spec,
+                                               String location, Map<String, String> properties) {
+    // create a new transaction without altering the cache. the table doesn't exist until the transaction is committed.
+    // if the table is created before the transaction commits, any cached version is correct and the transaction create
+    // will fail. if the transaction commits before another create, then the cache will be empty.
+    return catalog.newCreateTableTransaction(ident, schema, spec, location, properties);
+  }
+
+  @Override
+  public Transaction newReplaceTableTransaction(TableIdentifier ident, Schema schema, PartitionSpec spec,
+                                                String location, Map<String, String> properties, boolean orCreate) {
+    // create a new transaction without altering the cache. the table doesn't change until the transaction is committed.
+    // when the transaction commits, invalidate the table in the cache if it is present.
+    return CommitCallbackTransaction.addCallback(
+        catalog.newReplaceTableTransaction(ident, schema, spec, location, properties, orCreate),
+        () -> tableCache.invalidate(ident));
+  }
+
+  @Override
+  public boolean dropTable(TableIdentifier ident, boolean purge) {
+    boolean dropped = catalog.dropTable(ident, false);
+    tableCache.invalidate(ident);
+    return dropped;
+  }
+
+  @Override
+  public void renameTable(TableIdentifier from, TableIdentifier to) {
+    catalog.renameTable(from, to);
+    tableCache.invalidate(from);
```

Comment:
No, because the table was probably instantiated using the `from` name. Rather than making this require an update to the table instance, it is easier to reload the table on the next access.